pkg/nsx/services/firewall.go

```diff
@@ -281,18 +381,242 @@ func (service *SecurityPolicyService) addOperatorIfNeeded(expressions *data.List
+func (service *SecurityPolicyService) updateExpressionsMatchLables(matchLabels map[string]string, memberType string, expressions *data.ListValue) {
 	for k, v := range *util.NormalizeLabels(&matchLabels) {
 		service.addOperatorIfNeeded(expressions, "AND")
 		expression := service.buildExpression(
-			"Condition", "SegmentPort",
+			"Condition", memberType,
 			fmt.Sprintf("%s|%s", k, v),
-			"Tag", "EQUALS",
+			"Tag", "EQUALS", "EQUALS",
 		)
 		expressions.Add(expression)
 	}
 }
 
+// NSX understand the multiple values w.r.t a key in a joined string manner
+// this function iterates over input matchExpressions LabelSelectorRequirement
+// with same operator and Key, and merges them into one and values to a joined string
+// e.g.
+// - {key: k1, operator: NotIn, values: [a1, a2, a3]}
+// - {key: k1, operator: NotIn, values: [a2, a3, a4]}
+//  => {key: k1, operator: NotIn, values: [a1, a2, a3, a4]}
+func (service *SecurityPolicyService) mergeSelectorMatchExpression(matchExpressions []metav1.LabelSelectorRequirement) *[]metav1.LabelSelectorRequirement {
+	mergedMatchExpressions := make([]metav1.LabelSelectorRequirement, 0)
+	var meregedSelector metav1.LabelSelectorRequirement
+	var labelSelectorMap = map[metav1.LabelSelectorOperator]map[string][]string{}
+
+	for _, d := range matchExpressions {
+		_, exists := labelSelectorMap[d.Operator]
+		if !exists {
+			labelSelectorMap[d.Operator] = map[string][]string{}
+		}
+		_, exists = labelSelectorMap[d.Operator][d.Key]
+
+		if !exists {
+			labelSelectorMap[d.Operator][d.Key] = append(labelSelectorMap[d.Operator][d.Key], d.Values...)
+		} else {
+			labelSelectorMap[d.Operator][d.Key] = append(labelSelectorMap[d.Operator][d.Key], d.Values...)
+			labelSelectorMap[d.Operator][d.Key] = util.RemoveDuplicateStr(labelSelectorMap[d.Operator][d.Key])
+		}
+	}
+
+	for key, value := range labelSelectorMap {
+		for subKey, subValue := range value {
+			meregedSelector.Values = subValue
+			meregedSelector.Operator = key
+			meregedSelector.Key = subKey
+			mergedMatchExpressions = append(mergedMatchExpressions, meregedSelector)
+		}
+	}
+
+	return &mergedMatchExpressions
+}
+
+// Given NSX currently doesn't support 'In' LabelSelector, to keep design simple,
+// only allow just one 'In' LabelSelector in matchExpressions with at most of five values in it.
+func (service *SecurityPolicyService) validateSelectorOpIn(matchLabels map[string]string, matchExpressions []metav1.LabelSelectorRequirement) (int, error) {
+	var mexprInOpCount = 0
+	var mexprInValueCount = 0
+	var err error = nil
+	var errorMsg string = ""
+
+	for _, expr := range matchExpressions {
+		if expr.Operator == metav1.LabelSelectorOpIn {
+			mexprInOpCount++
+			mexprInValueCount += len(expr.Values)
+			if mexprInOpCount > MaxMatchExpressionInOp {
+				errorMsg = fmt.Sprintf("Count of Opeator IN expressions: %d exceed limit of %d",
```

Comment:
typo: `Opeator` -> `operator`
And I think here we need to use the lower case. Currently, the output looks like:
`Reason:   Error occurred while processing the Security Policy CR. Please check the config and try again. Error: Count of Operator IN expressions: 2 exceed limit of 1`

Reply:
And you interrupt the check loop... Then in the negative case, it will never report "**2** exceed 1", won't be "**3/4/5/6**... exceed 1", then printing `mexprInOpCount` won't be useful, right?
My suggestion is still that you check the total number in the end, i.e. check the number after the for loop(line 458).

Reply:
Thanks for suggestion. I've updated code as suggested.